crates/starknet-server/tests/common/mod.rs

```diff
@@ -1,53 +1,78 @@
-use std::net::TcpListener;
-use std::process::{Child, Command, Stdio};
-use std::{thread, time};
-
-use hyper::client::HttpConnector;
-use hyper::{Client, StatusCode, Uri};
-use thiserror::Error;
-
-#[derive(Error, Debug)]
-pub enum TestError {
-    #[error("Cannot start Devnet")]
-    DevnetNotStartable,
-    #[error("No free ports")]
-    NoFreePorts,
+#[cfg(test)]
+pub mod constants {
+    pub const HOST: &str = "localhost";
+    pub const MIN_PORT: u16 = 1025;
+    pub const MAX_PORT: u16 = 65_535;
+    pub const SEED: usize = 42;
+    pub const ACCOUNTS: usize = 3;
+
+    pub const MAX_RETRIES: usize = 10;
+
+    // predeployed account info with seed=42
+    pub const PREDEPLOYED_ACCOUNT_ADDRESS: &str =
+        "0x34ba56f92265f0868c57d3fe72ecab144fc96f97954bbbc4252cef8e8a979ba";
+    pub const EXPECTED_ACCOUNT_CLASS_HASH: &str =
+        "0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f";
 }
 
-const HOST: &str = "localhost";
-const MIN_PORT: u16 = 1025;
-const MAX_PORT: u16 = 65_535;
-const SEED: usize = 42;
-const ACCOUNTS: usize = 3;
+#[cfg(test)]
+pub mod util {
+    use std::net::TcpListener;
+    use std::process::{Child, Command, Stdio};
+    use std::{thread, time};
+
+    use hyper::{Client, StatusCode, Uri};
+    use lazy_static::lazy_static;
+    use starknet_rs_providers::jsonrpc::HttpTransport;
+    use starknet_rs_providers::JsonRpcClient;
+    use thiserror::Error;
+    use tokio::sync::Mutex;
+    use url::Url;
 
-const MAX_RETRIES: usize = 10;
+    use super::constants::{ACCOUNTS, HOST, MAX_PORT, MAX_RETRIES, MIN_PORT, SEED};
+
+    #[derive(Error, Debug)]
+    pub enum TestError {
+        #[error("No free ports")]
+        NoFreePorts,
+    }
 
-fn get_free_port_listener() -> Result<u16, TestError> {
-    for port in MIN_PORT..=MAX_PORT {
-        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
-            return Ok(listener.local_addr().expect("No local addr").port());
+    fn get_free_port_listener() -> Result<u16, TestError> {
+        for port in MIN_PORT..=MAX_PORT {
+            if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
+                return Ok(listener.local_addr().expect("No local addr").port());
+            }
+            // otherwise port is occupied
         }
-        // otherwise port is occupied
+        Err(TestError::NoFreePorts)
     }
-    Err(TestError::NoFreePorts)
-}
 
-pub(crate) struct BackgroundDevnet {
-    client: Client<HttpConnector>,
-    process: Option<Child>,
-}
+    lazy_static! {
+        /// This is to prevent TOCTOU errors; i.e. one background devnet might find one
```

Comment:
I think the mutex can be then locked only in `get_free_port_listener` and not for the whole `BackgroundDevnet::spawn`?

Reply:
It might be safer, but still not completely safe: one BackgroundDevnet can find a free port, release the mutex, and while it's trying to acquire it, another BackgroundDevnet can find the same free port. I can also add a comment about that.

Reply:
Added comment in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/64/commits/56701d07e723e0f0376393ee17caeda0e1f25b86

I also accidentally commited the removal of _mutex guard, but restored it in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/64/commits/748e6d07f014873ed72c0c8cb02e1eeb9bf8b17f